### PR TITLE
table partition: fix CREATE partitioned table fails check for unsigned int column definition

### LINKED ISSUE #35827: CREATE partitioned table fails check for unsigned int column definition
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table t (id tinyint unsigned, idpart tinyint, i varchar(255)) partition by range (idpart) (partition p1 values less than (-1));
```

### 2. What did you expect to see? (Required)
```
mysql> create table t (id tinyint unsigned, idpart tinyint, i varchar(255)) partition by range (idpart) (partition p1 values less than (-1));
Query OK, 0 rows affected (0,03 sec)
```

### 3. What did you see instead (Required)
```
tidb> create table t (id tinyint unsigned, idpart tinyint, i varchar(255)) partition by range (idpart) (partition p1 values less than (-1));
ERROR 1563 (HY000): Partition constant is out of partition function domain
```

### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v6.2.0-alpha-230-g47a4f3a017
Edition: Community
Git Commit Hash: 47a4f3a017768dda0b9a826451486c5539bc8427
Git Branch: master
UTC Build Time: 2022-06-29 19:09:58
GoVersion: go1.18
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
Store: unistore
```

### PR DESCRIPTION
**Pull request by ymkzpx**
### What problem does this PR solve?

Issue Number: close #35827

Problem Summary:

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: ddl/partition.go [hunk 1]
```diff
@@ -660,9 +660,11 @@ func buildRangePartitionDefinitions(ctx sessionctx.Context, defs []*ast.Partitio
 
 func checkPartitionValuesIsInt(ctx sessionctx.Context, def *ast.PartitionDefinition, exprs []ast.ExprNode, tbInfo *model.TableInfo) error {
 	tp := types.NewFieldType(mysql.TypeLonglong)
-	if isColUnsigned(tbInfo.Columns, tbInfo.Partition) {
+	col := model.FindColumnInfo(tbInfo.Columns, tbInfo.Partition.Expr[1:len(tbInfo.Partition.Expr)-1])
```
Comment (ymkzpx) [inline ddl/partition.go:663]: `tbInfo.Partition.Expr` is a string with single quotes，so `tbInfo.Partition.Expr[1:len(tbInfo.Partition.Expr)-1]`.

**Comment by ymkzpx** [comment]
@mjonss If the contain is changed to equal, the situation cannot be handled, unit-test will fail, as shown below.
image

**Comment by mjonss** [comment]
> @mjonss If the contain is changed to equal, the case can't be handled, unit-test will fail, as shown below. image

So it will get another error? What would you suggest to handle this, accept the new error or fix the code to retain the old error?

**Comment by ymkzpx** [comment]
> So it will get another error? What would you suggest to handle this, accept the new error or fix the code to retain the old error?

@mjonss
I checked the code and found that if expression is less than 0, TiDB currently does not check for unsigned data types, which may cause bugs, so i updated the code to fix this, PTAL.

**Review by mjonss — COMMENTED** [review]
LGTM, just two minor things to address.

File: ddl/partition.go [hunk 2]
```diff
@@ -660,9 +660,11 @@ func buildRangePartitionDefinitions(ctx sessionctx.Context, defs []*ast.Partitio
 
 func checkPartitionValuesIsInt(ctx sessionctx.Context, def *ast.PartitionDefinition, exprs []ast.ExprNode, tbInfo *model.TableInfo) error {
 	tp := types.NewFieldType(mysql.TypeLonglong)
-	if isColUnsigned(tbInfo.Columns, tbInfo.Partition) {
```
[the remark this answers is not in the document]
Comment (ymkzpx) [reply ddl/partition.go:663]: Currently, TiDB has a problem recognizing unsigned data type. If we change the condition substring to equal, for example, an expression like `id-10`, id is unsigned data type, TiDB can't handle.

File: errno/errcode.go [hunk 3]
```diff
@@ -822,6 +822,7 @@ const (
 	ErrMustChangePasswordLogin                               = 1862
 	ErrRowInWrongPartition                                   = 1863
 	ErrErrorLast                                             = 1863
+	ErrUnsignedNotLessThanZero                               = 1865
```
Comment (mjonss) [inline errno/errcode.go:825]:
```suggestion
	ErrUnsignedLessThanZero                                  = 1865
```

File: ddl/partition.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mjonss) [reply ddl/partition.go:663]: Any suggestions on how to solve this? I still think we should not keep the broken isColUnsigned, we should either fix it or remove it ;)
In some edge cases it is OK to not be 100% compatible with MySQL if we have a good reason not to.

**Comment by mjonss** [comment]
Maybe something like [this](https://github.com/pingcap/tidb/pull/36830)?

**Pull request closed without merging**